hax-lib-macros/src/lib.rs

```diff
@@ -1,93 +1,363 @@
-const HAX_TOOL: &str = "_hax";
-
-macro_rules! no_argument {
-    () => {
-        |at: TokenStream2| {
-            if !at.is_empty() {
-                abort!(at, "this attribute doesn't take any argument")
-            }
-            at
-        }
-    };
+#![feature(let_chains)]
+#![feature(box_patterns)]
+
+mod rewrite_self;
+mod syn_ext;
+mod utils;
+
+mod prelude {
+    pub use proc_macro as pm;
+    pub use proc_macro2::*;
+    pub use proc_macro_error::*;
+    pub use quote::*;
+    pub use syn::spanned::Spanned;
+    pub use syn::{visit_mut::VisitMut, *};
+
+    pub use hax_lib_macros_types::*;
+    pub use AttrPayload::Language as AttrHaxLang;
+    pub type FnLike = syn::ImplItemFn;
 }
 
-/// `passthrough_attribute!(NAME)` generates a proc-macro that expands
-/// into the tool attribute `HAX_TOOL::NAME` when the cfg flag
-/// `HAX_COMPILATION` is set.
-macro_rules! passthrough_attribute {
-    ($(#$a:tt)*$name:ident) => {
-        passthrough_attribute!($(#$a)*$name, no_argument!());
-    };
-    ($(#$a:tt)*$name:ident, |$x:pat_param| $e:expr) => {
-        passthrough_attribute!($(#$a)*$name, |$x: TokenStream2| $e);
-    };
-    ($(#$a:tt)*$name:ident, $validator:expr) => {
-        #[proc_macro_error]
-        #[proc_macro_attribute]
-        $(#$a)*
-        pub fn $name(attr: TokenStream, item: TokenStream) -> TokenStream {
-            let attr: TokenStream2 = attr.into();
-            let item: TokenStream2 = item.into();
-            let hax_compilation = format_ident!("{}", HAX_COMPILATION);
-            let hax_tool = format_ident!("{}", HAX_TOOL);
-            let attr: TokenStream2 = $validator(attr);
-            quote! {
-                #[cfg_attr(#hax_compilation, #hax_tool::$name(#attr))]
-                #item
-            }
-            .into()
-        }
+use prelude::*;
+use syn_ext::*;
+use utils::*;
+
+/// Include this item in the Hax translation.
+#[proc_macro_error]
+#[proc_macro_attribute]
+pub fn include(attr: pm::TokenStream, item: pm::TokenStream) -> pm::TokenStream {
+    let item: TokenStream = item.into();
+    let _ = parse_macro_input!(attr as parse::Nothing);
+    let attr = AttrPayload::ItemStatus(ItemStatus::Included { late_skip: false });
+    quote! {#attr #item}.into()
+}
+
+/// Exclude this item from the Hax translation.
+#[proc_macro_error]
+#[proc_macro_attribute]
+pub fn exclude(attr: pm::TokenStream, item: pm::TokenStream) -> pm::TokenStream {
+    let item: TokenStream = item.into();
+    let _ = parse_macro_input!(attr as parse::Nothing);
+    let attr = AttrPayload::ItemStatus(ItemStatus::Excluded { modeled_by: None });
+    quote! {#attr #item}.into()
+}
+
+/// Exclude this item from the Hax translation, and replace it with a
+/// axiomatized model in each backends. The path of the axiomatized
+/// model should be given in Rust syntax.
+///
+/// # Example
+///
+/// ```
+/// use hax_lib_macros::*;
+/// #[modeled_by(FStar::IO::debug_print_string)]
```

Comment:
How is this path interpreted?
And can I add a second one with `Coq::A::b`?

Reply:
there is no treatments in the backend yet, so that's not defined yet.
I'm disabling this `modeled_by` attribute for now thus